WIP: Resolve Cabal keys using DNS

MarkBennett opened the pull request:
This is a work in progress but I'm looking for some feedback @noffle. I'm not sensitive, so any feedback is welcome.

# Questions

* What `nodejs` version are we targetting? Can I use ES2015+ (`class`, `async/await`, `arrow functions`)
* I'll probably need to do some more complex mocking to test DNS caching. Prefs on mocking framework?
* How do you feel about adding the `Cabal#close()`? I need this if the cache will automatically expire old values, but realize having to call `close()` might be annoying and require refactoring other code. Would you prefer I use a cache that doesn't automatically expire keys and only drops them on `get()`? Doing this could potentially lead to memory issues if the cache somehow fills up, but that is probably not a huge issue at this point.
* Can we setup a `cabalkey=cabal://4ae5ec168a9f6b45b9d35e3cc1d0f4e3a436000d37fae8f53b3f8dadfe8f192f` TXT on `cabal.chat` that I can use in the end-to-end testing instead of `markbennett.ca`?

# TODO:
* [x] Store lookups in cache and test that this works
* [x] Refactor `resolve()` to work with an array of resolvers as there's lots of duplicated code
* [x] Handle DNS lookup errors (network, JSON formatting, etc.)
* [x] Generally abuse params and cleanup error handling

Closes #7

noffle commented on resolve.js:
```diff
@@ -0,0 +1,123 @@
+const assert = require('assert')
```
unused

noffle commented on index.js:
```diff
@@ -1,9 +1,20 @@
 var through = require('through2')
+var Resolver = require('./resolve')
+
+var JSON_VALUE_ENCODING = {
```
:+1:

noffle commented on resolve.js:
```diff
@@ -0,0 +1,123 @@
+  return key
+}
+
+function dnsResolver(hostname, cb) {
```
Could we use node's `dns` module here?  e.g.

```
$ cat > dns.js
var dns = require('dns')

dns.resolveTxt('beakerbrowser.com', console.log)
^D

$ node dns.js 
null [ [ 'DATKEY=87ed2e3b160f261a032af03921a3bd09227d0a4cde73466c17114816cae43336' ] ]
```

MarkBennett replied:
I noticed that `dat-dns` chose to use DNS-over-HTTPS as they believed that it offered more confidential lookups, hence using it here. I'm fine with using the node dns if you think that's ok. It's written so that the dns resolver can be easily replaced if someone had concerns one way or the other.

Just let me know what DNS you want and I'll wire it up!

noffle replied:
Can you say a bit more about this, or link me to what the dat folks said?

Is the idea that hitting your ISP for DNS (probably most people's default) is less desirable than google's servers?

MarkBennett replied:
Sure. I can't remember exactly where I saw it but I'll track it down!

MarkBennett replied:
I've dropped the Google DNS and am just using whatever's configured on the system.

Along with this change, I've also dropped the caching and gotten rid of `Cabal.close()` as well.

noffle commented on resolve.js:
```diff
@@ -0,0 +1,123 @@
+const assert = require('assert')
+const encoding = require('dat-encoding')
+const https = require('https')
+const NodeCache = require( "node-cache" )
```
This is just an in-memory cache, right? Nothing persisted to disk?

What's the use case for an in-memory cache? When I join a cabal, my cabal-core instance will resolve a cabal URL exactly once, right?

noffle replied:
(Once we start building out cabal-client, I think that will be a great place to maintain an on-disk cache of resolved cabal URLs.)

MarkBennett replied:
> This is just an in-memory cache, right? Nothing persisted to disk?
> 
> What's the use case for an in-memory cache? When I join a cabal, my cabal-core instance will resolve a cabal URL exactly once, right?

Yes. This is supposed to just be in-memory but I'll double check to be sure.

Do you think we should just take the caching out until we know how often it's used?

noffle replied:
+1 on not adding in-memory caching until there's a need for it. This also removes the need for a `close` API.

MarkBennett commented:
> Could you add a note in the README somewhere that mentions that cabal-core can resolve via DNS now?

Good catch. I'll add it to the `README`.

MarkBennett commented:
> Have you tested this with the cabal cli? I'm surprised that the frontend doesn't complain about the db not being available.

👍  I haven't build the cli against this yet, I'll link it up and check it out!

MarkBennett commented:
Thanks for the review @noffle. I'll address your comments over the next couple of days.

Any thoughts on `async/await`, `Cabal#close()`?

MarkBennett commented:
> I set a TXT entry on cabal.chat to cabalkey=4ae5ec168a9f6b45b9d35e3cc1d0f4e3a436000d37fae8f53b3f8dadfe8f192f

🎉  I'll update the tests. Thanks! :)

noffle commented:
@MarkBennett re async/await/etc, I like your current style (not using them). I mostly write ES5.

noffle commented:
I don't like the idea of unit tests that rely on an internet connection. Having a mocking setup or something would be cool. :+1:   Another option is having another testing section explicitly for integration tests. What do you think?

MarkBennett commented:
> I don't like the idea of unit tests that rely on an internet connection. Having a mocking setup or something would be cool. 👍 Another option is having another testing section explicitly for integration tests. What do you think?

Sure. I like separating those out into unit tests that don't depend on the network, and a seperate integration file that works end-to-end. I'll go ahead and do that with a library that mocks out the node network API.

MarkBennett commented:
@noffle I found out why DAT resolves with Google DNS. Apparently normal UDP happens over unencrypted UDP so you can't verify that the DNS lookup is unmodified or intercepted by third parties.

https://www.datprotocol.com/deps/0005-dns/

That said if users are concerned about DNS lookup security this would be an issue for all DNS lookups on their system, so they could configure their system DNS to meet their own privacy needs.

I've also realised that user Google DNS also makes it impossible for me to run my own DNS server and resolve lookups on the local network.

Unless you see a strong reason not to, I'm going to switch to the Node DNS instead.

noffle commented:
+1. Thanks for doing the research.

MarkBennett commented:
> I set a TXT entry on cabal.chat to cabalkey=4ae5ec168a9f6b45b9d35e3cc1d0f4e3a436000d37fae8f53b3f8dadfe8f192f

Can you confirm that this TXT on cabal.chat is working? This is what I get back from `dig`:

~~~
22:30 ~/projects/cabal-core (#7-resolve-cabals-with-dns)
$ dig q-type TXT cabal.chat

; <<>> DiG 9.10.6 <<>> q-type TXT cabal.chat
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 20870
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;q-type.				IN	TXT

;; AUTHORITY SECTION:
.			86370	IN	SOA	a.root-servers.net. nstld.verisign-grs.com. 2018082701 1800 900 604800 86400

;; Query time: 38 msec
;; SERVER: 192.168.1.1#53(192.168.1.1)
;; WHEN: Mon Aug 27 22:30:41 MDT 2018
;; MSG SIZE  rcvd: 110

;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 55257
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;cabal.chat.			IN	A

;; ANSWER SECTION:
cabal.chat.		1299	IN	A	192.64.119.129

;; Query time: 2 msec
;; SERVER: 192.168.1.1#53(192.168.1.1)
;; WHEN: Mon Aug 27 22:30:41 MDT 2018
;; MSG SIZE  rcvd: 55
~~~

MarkBennett commented:
@noffle So I think I've handled most of the feedback so far. I have found an interesting issue though...

In the cabal cli, and I'm assuming other clients, they operate under the assumption that the cabal is ready to go as soon an initialized, then they wait for the db to initialize before carrying on. ([See here](https://github.com/cabal-club/cabal/blob/master/cli.js#L40))

~~~js
var cabal = Cabal(args.db, args.key, {username: nick})
cabal.db.on('ready', function () {
  if (!args.seeder) {
    frontend(cabal)
  } else {
    console.log('reseeding the cabal at cabal://' + cabal.db.key.toString('hex'))
  }
  cabalSwarm(cabal)
})
~~~

This doesn't work anymore though as the `db` isn't even initialized until the key is known which doesn't happen until the callback to the new `resolve()` function is triggered.

To work around this I see a few options:

1) Emit a `ready` event from the cabal when it's done resolving the key and has initialized the database
2) Pass a callback to the cabal constructor which in invoked after the constructor is complete

As the cabal is already an `EventEmitter`, adding a new `ready` event is my preferred solution (option 1) but it will require some refactoring of the CLI and any other clients using cabal. I'm happy to commit this for the CLI but haven't looked the desktop code enough to do that.

Let me know what you prefer and I'll get to work!

noffle commented:
Whoops, I think I did the TXT entry wrong. Tried again. Let me know if you're getting anything different.

noffle commented:
Another approach I was thinking about was having resolution be a "static method", so cabal cli could do:

```js
Cabal.resolveKey(args.key, function (err, key) {
  if (err) throw err
  var cabal = Cabal(args.db, key, {username: nick})
  cabal.db.on('ready', function () {
    if (!args.seeder) {
      frontend(cabal)
    } else {
      console.log('reseeding the cabal at cabal://' + cabal.db.key.toString('hex'))
    }
    cabalSwarm(cabal)
  })
})
```
This lets us skip a lot of the timing & error reporting complexities that come up in trying to cram the resolver into the constructor / db init.

MarkBennett commented:
> Another approach I was thinking about was having resolution be a "static method"

Interesting idea @noffle. I like that it keeps things simpler in the constructor, but it feels like adding the static method might have other issues.

Without raining on this too much, here are some thoughts...

For one, how do the client apps determine when to use the static resolver vs just passing in the key? Would it run a regex to check for a dat key then determine if it needs to run a `resolveKey()` first? Would it stipulate that the app only accepts keys. Given that the resolve method also handles sanitizing the many ways of entering the key (cabal:<key>, dat:<key>, <key>, etc.) there would probably need to be some code either in the constructor or the client to validate the provided key which would duplicate what's in the `resolveKey()` function already.

The more I look at this code the more I think reaching into `cabal.db` to add the `on('ready')` handler ties the code to whatever `db` implementation `cabal-core` uses. Regardless of how the resolver works, it seems like it would be better to have a `cabal.on(ready)` instead so that the underlying implementation is abstracted away. This issue with waiting to resolve the key only highlights this problem of being tied to the underlying `db` being ready rather than the `cabal` itself.

I'm thinking the refactored client code that looks like this:

~~~js
var cabal = Cabal(args.db, args.key, {username: nick})
cabal.on('ready', function () {
  if (!args.seeder) {
    frontend(cabal)
  } else {
    console.log('reseeding the cabal at cabal://' + cabal.db.key.toString('hex'))
  }
  cabalSwarm(cabal)
})
~~~

Basically, we should just need to refactor the current `cabal.db.on('ready')` call to `cabal.on('ready')` instead.

I hope I'm explaining myself OK and not just stomping on your idea. I really do appreciate the input so if you see something wrong with this idea that I'm missing, please let me know!

MarkBennett commented:
> This lets us skip a lot of the timing & error reporting complexities that come up in trying to cram the resolver into the constructor / db init.

One other thought just occurred to me... what about using the constructor to just setup the cabal configuration, then using an `init()` method to prepare it? In the code above is there any guarantee that `cabal.on('ready')` will run before the callbacks in the constructor? Using an `init()` method after the handlers are wired up would guarantee that the app is ready to handle events when the initialization finishes.

noffle commented:
I'm down with having a `cabal.on('ready')` and using that. (Let's also have a `cabal.on('error')` too, for emitting resolution or db init problems)

cblgh commented:
+1

On Wed, 29 Aug 2018 at 19:51, Stephen Whitmore <notifications@github.com>
wrote:

> I'm down with having a cabal.on('ready') and using that. (Let's also have
> a cabal.on('error') too, for emitting resolution or db init problems)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cabal-club/cabal-core/pull/13#issuecomment-417044782>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADrvWmfiw6vO2Ug-4nCwGzUKXYLQAduJks5uVtS5gaJpZM4WLp_k>
> .
>

karissa commented:
There's already a module for this, https://www.npmjs.com/package/dat-link-resolve and https://www.npmjs.com/package/dat-dns

MarkBennett commented:
> There's already a module for this, https://www.npmjs.com/package/dat-link-resolve and https://www.npmjs.com/package/dat-dns

@karissa thanks for the heads up on those. We actually looked at them and much of this code is based off what's in there, however it operates a little differently as `cabal` actually uses `hypercore` which sits under `dat` rather than using it directly. That means some of the behaviors in `dat-dns` don't make sense for `cabal`.

There's some discussion of it in #7.

karissa commented:
@mark dat links are also hypercore links. The only differences I see are the custom “cabalkey” and “cabal://“ prefixes, but supporting this use case would be a nice feature to have in dat-dns or dat-link-resolve. You could add a `prefixes` option that expandan the search space beyond just `dat`

I recommend re using modules whenever possible here. it’s possible that someone will want to publish a cabal-dat using dat prefixes on hashbase, for example.

dat-dns and dat-link-resolve are used in beaker and fritter and the dat cli, so cabal will benefit from that cross compatibility in the ecosystem.

MarkBennett commented:
Thanks @karissa. Interesting. I've based most of the code in this PR off that so it woudn't be a huge refactor.  The only thing that might not make sense is the way `dat` resolves things with the `.well-known/dat` file. I'm not sure re-using the `.well-known/dat` file would make sense for `cabal`. So there might need to be a way to turn that off. We also chose to use the system DNS rather than relying on Google as there might be cases where you'd want to be able to resolve DNS entries locally without reaching out to Google first. So adding a resolver able to do that to the `dat` packages might make sense.

Thoughts? What do you think @noffle?

I'm going to have some more time to look at this PR this weekend so let me know and I'll integrate feedback.

MarkBennett commented:
@noffle @karissa I've been thinking about this over lunch. To keep this PR moving along, I'm thinking I'd like to finish it off as is with DNS resolution working on it's own in `cabal`.

Then I can contact `dat-dns` and `dat-link-resolve` and file issues to see if and how they'd like to integrate cabal hypercore key resolution into `dat`.

This will also give us a chance to see how `cabal` DNS resolution is used and provide some evidence if there's enough overlap to integrate with `dat`.

noffle commented:
Sounds good to me. Anything left to do on this @MarkBennett?

MarkBennett commented:
> Sounds good to me. Anything left to do on this @MarkBennett?

@noffle I need to refactor to use the events still and then do some more testing. I'm slammed right now that a projeft and a talk in DAT on Thursday. Guessing I'll wrap it be week? Is it blocking stuff? If so I'll try to wrap it faster.

noffle commented:
Nope, you're good! But we're going to have to rebase this onto the new kappa core branch (`refactor`) at some point. :cry:

noffle commented:
Non-blocking! Just a super cool feature to have. :)

If it's done and merged into the kappa stuff before this Friday (somehow), we
can advertise `cabal://cabal.chat` as the public key to join! But seriously, no
pressure. Take care of yourself. :green_heart:


On 09/01  09:17, Mark Bennett wrote:
> > Sounds good to me. Anything left to do on this @MarkBennett?
> 
> @noffle I need to refactor to use the events still and then do some more testing. I'm slammed right now that a projeft and a talk in DAT on Thursday. Guessing I'll wrap it be week? Is it blocking stuff? If so I'll try to wrap it faster.
> 
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/cabal-club/cabal-core/pull/13#issuecomment-417870077

noffle commented:
Update: woo, I see the `cabalkey` showing up as a TXT entry now :sunglasses:

noffle commented:
@MarkBennett I sure am!

MarkBennett commented:
@noffle It looks like some of the kappa changes have been merged into `master` already. Is the `kappa` code in `master` now or in the `refactor` branch?

noffle commented:
Yep, it's all in master now.

MarkBennett commented:
@noffle I've updated this branch with `master` and `kappa` now, however I wanted to get your feedback as I've introduced a `cabal#ready` event and a `cabal#init()` method to ensure `ready` only fires after you've had a chance to wire up your listeners.

Thoughts?

MarkBennett commented:
Hey @noffle. I know you're probably busy so I don't want to bug you. Just wanted to make sure you saw my last message.

Let me know if you're good with the `cabal#ready` and `cabal#error` events, as well as `cabal#init()`.

If you are we should talk about how to update the other cabal clients and maybe get some feedback from them.

noffle commented on README.md:
```diff
@@ -4,101 +4,66 @@ Core database, replication, and chat APIs for cabal.
 ## API
 
-> var Cabal = require('cabal-node')
+<!-- Generated by documentation.js. Update this documentation by updating the source code. -->
```
sorry could you remove this auto-generated documentation? looking at the diff I'm seeing useful information being removed from the readme. maybe we could discuss the idea of auto-docs in another PR?

MarkBennett replied:
I was just running the `npm run doc` command. I'll update the `README.md` from master then run it again and see if this fixes it. Otherwise, I might need a pointer if I'm still stuck. Thanks!

noffle commented:
Hey @MarkBennett I know this PR has maybe been more back-and-forth work than expected, but I appreciate your sticking with it and working to get this cool feature landed :star2:

MarkBennett commented:
> var cabal = new Cabal(blah, blah) cabal.ready(function (err) { // ... })

Sounds good. If you call this multiple times will it just immediately invoke the callback if the Cabal is already `ready`?

MarkBennett commented:
@noffle  I'm going to need to take a break from working on this for a few weeks. Not for lack of enthusiasm, but I've been putting off doing my year end taxes and now they're due. I'm also spinning up a not-for-profit and promised the other members I'd give it some love.

I'll try to sneak in some commits over the next few weeks, but you might notice I'm a little slower. Sorry, and feel free to fork if you're anxious to get this in sooner.

You rock! 🎉

cblgh commented:
@noffle @MarkBennett this looks pretty close to done, if not done already right? 

can we merge it into master so that i can make use of `cabal.ready(cb)` in https://github.com/cabal-club/cabal-core/pull/22?

noffle commented:
@cblgh I still see some outstanding todos:

- fix branch merge conflicts
- revert the readme changes (since it deletes lots of good useful content) & update to reflect mark's API changes
- change to using `cabal-core#ready` instead of event emitters
- phase out `cabal-core#init` and have the constructor do it automatically

MarkBennett commented:
Remaining todos look right to me @noffle.

Sorry @cblgh! :(

MarkBennett commented:
Hey! Just a heads up that I haven't forgotten about this. I will get back to it, just had some laptop trouble then had to bump some priorities and couldn't push this up above them. Sorry for the delay!

noffle commented:
Hey Mark no worries! Take care of yo-self & take your time.

cblgh commented:
we actually have this now by way upstream changes i made in [dat-dns](https://github.com/datprotocol/dat-dns/pull/15), closing this

but thank you for your effort in getting this started @MarkBennett!

MarkBennett commented:
Sorry I was such a flake on this @cblgh and thanks for being so gracious about it. ❤️ 

Awesome to hear that the changes are still incoming! 👍
